Assets/Scripts/GameController.cs

```diff
@@ -48,105 +52,238 @@ public class GameController : MonoBehaviour {
+		EvalSurroundingTokenActions(unit, token, range, "move");
 
+		// After all actions have been assessed, paint the tokens
 		foreach(ValidAction action in Actions) {
-			Tokens[action.col][action.row].SetActionProperties(action.action);
+			Tokens[action.col][action.row].PaintAction(action.action);
 		}
 	}
 
-	private void EvalSurroundingTokenActions(string unitName, float range, Token token) {
+	// Evaluates the surrounding tokens for valid actions
+	private void EvalSurroundingTokenActions(Unit unit, Token token, float range, string action) {
 		Token nextToken;
 		// Start with above token
 		if(token.Y < GridHeight) {
 			nextToken = Tokens[token.X][token.Y+1];
-			EvalNextToken(unitName, range, nextToken);
+			EvalNextToken(unit, nextToken, token, range, action);
 		}
 		// Start with left token
 		if(token.X > 0) {
 			nextToken = Tokens[token.X-1][token.Y];
-			EvalNextToken(unitName, range, nextToken);
+			EvalNextToken(unit, nextToken, token, range, action);
 		}
 		// Start with below token
 		if(token.Y > 0) {
 			nextToken = Tokens[token.X][token.Y-1];
-			EvalNextToken(unitName, range, nextToken);
+			EvalNextToken(unit, nextToken, token, range, action);
 		}
 		// Start with right token
 		if(token.X < GridLength) {
 			nextToken = Tokens[token.X+1][token.Y];
-			EvalNextToken(unitName, range, nextToken);
+			EvalNextToken(unit, nextToken, token, range, action);
 		}
 	}
 
-	private void EvalNextToken(string unitName, float range, Token token) {
-		if(!token.CanMove && !token.CanAttack) {
-			float remainingRange;
-			remainingRange = EvalToken(unitName, range, token);
-			if(remainingRange > 0f) {
-				EvalSurroundingTokenActions(unitName, remainingRange, token);
+	// Evaluates the specified token and determines to either continue movement, switch to attack, or end
+	private void EvalNextToken(Unit unit, Token token, Token prevToken, float range, string action) {
+		// Evaluate this token for valid actions, and calculate remaining range, if any
+		float remainingRange;
+		remainingRange = EvalToken(unit.name, token, range, action);
+		if(remainingRange >= 0f) {
+			// If there is range remaining, evaluate surrounding tokens with whatever the current action is (move or attack)
+			EvalSurroundingTokenActions(unit, token, remainingRange, action);
+		}else if(action == "move") {
+			// If there is no range remaining (i.e. the unit can NOT move into this token)
+			// and the action is currently "move"
+			if(prevToken.CurrentUnit != null) {
+				// We can't move here (remainingrange is less than 0)
+				// We also can't move to the previous token because prevToken.CurrentUnit != null
+				// So we need to evaluate the tokens surrounding the PREVIOUS token
+				// Use AtkRange-1 because the previous token uses the first tick of attack range
+				// If we saved the previous previous token, we could eval surrounding with AtkRange
+				EvalSurroundingTokenActions(unit, prevToken, unit.GetStat("AttackRange").Value-1, "attack");
```

Comment:
I would just see if you can make it more concise.  For instance, "// We can't move here (remainingrange is less than 0)", I can deduce that based on the if statement.  It didn't hit the >= 0 part so maybe that part doesn't need its own dedicated comment here.  Just talking aloud.